regcomp.c

```diff
@@ -19170,18 +19170,130 @@ S_regclass(pTHX_ RExC_state_t *pRExC_state, I32 *flagp, U32 depth,
         RExC_contains_locale = 1;
     }
 
-    /* Some character classes are equivalent to other nodes.  Such nodes take
-     * up less room, and some nodes require fewer operations to execute, than
-     * ANYOF nodes.  EXACTish nodes may be joinable with adjacent nodes to
-     * improve efficiency. */
-
     if (optimizable) {
+
+        /* Some character classes are equivalent to other nodes.  Such nodes
+         * take up less room, and some nodes require fewer operations to
+         * execute, than ANYOF nodes.  EXACTish nodes may be joinable with
+         * adjacent nodes to improve efficiency. */
+        op = optimize_regclass(pRExC_state, cp_list,
+                                            only_utf8_locale_list,
+                                            upper_latin1_only_utf8_matches,
+                                            has_runtime_dependency,
+                                            posixl,
+                                            &anyof_flags, &invert, &ret, flagp);
+        RETURN_FAIL_ON_RESTART_FLAGP(flagp);
+
+        /* If optimized to something else, finish up and return */
+        if (ret >= 0) {
+            Set_Node_Offset_Length(REGNODE_p(ret), orig_parse - RExC_start,
+                                                   RExC_parse - orig_parse);;
+            SvREFCNT_dec(cp_list);;
+            SvREFCNT_dec(only_utf8_locale_list);
+            SvREFCNT_dec(upper_latin1_only_utf8_matches);
+            return ret;
+        }
+    }
+
+    /* Here didn't optimize, or optimized to a specialized ANYOF node.  If the
+     * former, set the particular type */
+    if (op == ANYOF) {
```

Comment:
It feels like this block could quite sanely be in the new function too - it appears already to have all the inputs required. That would leave the new function making all the decisions about which op to emit, which would make reasoning about all this slightly easier. The only downside I see is that it probably involves bringing back the "is_anyof" label.

Reply:
No, unless I misunderstand what you mean.  Calling the new function is skipped when the node cannot be optimized to something besides an ANYOF node; which happens for various reasons.  So deciding which ANYOF node to use has to be outside the block.

Reply:
Ah right, fair point.

Reply:
Let me rephrase that.  It is probably true that the new function has the required inputs to make this decision.  But that new function is sometimes skipped, so the decision has to be made as well when the function isn't involved.  Better to not repeat the logic.